src/Compilers/CSharp/Test/Symbol/Symbols/FunctionPointerTypeSymbolTests.cs

```diff
@@ -0,0 +1,611 @@
+                Equality.Equal, Equality.Equal, Equality.Equal, Equality.Equal);
+        }
+
+        [Fact]
+        public void Equality_ParameterModifiersDifferent()
+        {
+            var comp = CreateCompilation(@"
+class C
+{
+    void M(delegate*<ref string, object, C, object, void> p1,
+           delegate*<string, in object, out C, object, void> p2) {}
+}", parseOptions: TestOptions.RegularPreview);
+
+            comp.VerifyDiagnostics();
+            var c = comp.GetTypeByMetadataName("C");
+            var m = c.GetMethod("M");
+            AssertEqualityAndHashCode((FunctionPointerTypeSymbol)m.Parameters[0].Type, (FunctionPointerTypeSymbol)m.Parameters[1].Type,
+                returnEquality: Equality.Equal, callingConventionEquality: Equality.Equal,
+                Equality.NotEqual, Equality.NotEqual, Equality.NotEqual, Equality.Equal);
+        }
+
+        [Fact]
+        public void Equality_ReturnTypeDifferent()
+        {
+            var comp = CreateCompilation(@"
+class C
+{
+    void M(delegate*<string, object, C, object, ref string> p1,
+           delegate*<string, object, C, object, ref readonly string> p2) {}
+}", parseOptions: TestOptions.RegularPreview);
+
+            comp.VerifyDiagnostics();
+            var c = comp.GetTypeByMetadataName("C");
+            var m = c.GetMethod("M");
+            AssertEqualityAndHashCode((FunctionPointerTypeSymbol)m.Parameters[0].Type, (FunctionPointerTypeSymbol)m.Parameters[1].Type,
+                returnEquality: Equality.NotEqual, callingConventionEquality: Equality.Equal,
+                Equality.Equal, Equality.Equal, Equality.Equal, Equality.Equal);
+        }
+
+        enum Equality
+        {
+            Equal = 0,
+            DifferingNullability = 0b01,
+            NotEqual = 0b10
+        }
+
+        private void AssertEqualityAndHashCode(FunctionPointerTypeSymbol p1, FunctionPointerTypeSymbol p2, Equality returnEquality, Equality callingConventionEquality, params Equality[] parameterEqualities)
+        {
+            var overallEquality = returnEquality | callingConventionEquality | (parameterEqualities.Length > 0 ? parameterEqualities.Aggregate((acc, cur) => acc | cur) : 0);
+
+            assertSymbolEquality(p1, p2, overallEquality);
+            assertSymbolEquality(p1.Signature, p2.Signature, overallEquality);
+
+            var ret1 = p1.Signature.ReturnTypeWithAnnotations;
+            var ret2 = p2.Signature.ReturnTypeWithAnnotations;
+            if (hasFlag(returnEquality, Equality.NotEqual))
+            {
+                if (p1.Signature.RefKind == p2.Signature.RefKind)
+                {
+                    Assert.False(ret1.Equals(ret2, TypeCompareKind.ConsiderEverything));
+                    Assert.False(ret1.Equals(ret2, TypeCompareKind.AllNullableIgnoreOptions));
+                    Assert.NotEqual(ret1.GetHashCode(), ret2.GetHashCode());
```

Comment:
>Assert.NotEqual(ret1.GetHashCode(), ret2.GetHashCode()); [](start = 20, length = 56)

It is incorrect to test HashCode for inequality. There is no requirement for HashCode be different. There is a requirement for HashCode be equal for equal objects.